test/drop_test/action.py

```diff
@@ -76,6 +78,215 @@ def compare_json_struct(expected: dict, actual: dict):
                     )
 
 
+def compare_value(expected: dict, typename: str, key: str, aval):
+    if not key in expected:
+        return
+
+    eval = expected[key]
+    if eval == "*":
+        return
+
+    if aval != eval:
+        raise Exception(f"{typename} `{key}` mismatch. Expected: {eval}, actual {aval}")
+
+
+def compare_json_to_incoming_path_state(
+    expected: dict, actual: norddrop.IncomingPathState
+):
+    keys = ["created_at", "state"]
+
+    compare_value(expected, "incomingPathState", "created_at", actual.created_at)
+
+    if actual.kind.is_pending():
+        compare_value(expected, "IncomingPathState", "state", "pending")
+        keys.append("base_dir")
```

Comment:
I do not trust this code(all these `compare_*` functions) :dagger: 
In `AssertTransfers` we encode the JSON argument - but we're typed very well - we could pass in exact typed data/objects we expect and comparison then would be free if each structure has `__hash__` defined which would be much nicer.

The testsuite is the thing which allows us to do such massive changes and trust the outputs but in this case I am not that sure. The code looks on the surface correct but the proposed method should make it more obvious and resilient imo. What do you think about this?

Reply:
So the thing is that we cannot use uniffy structures directly because `__eq__` compares all the fields and for example `created_at` we don't want to compare. This leaves us with two options:
* custom types and then mapping into uniffy types - this is the best approach probably, but is a ton of work (`scenarios.py` needs to be rewritten in terms of new structures)
* reuse the current JSON format and compare it with the uniffy structs - that's the thing I did in this PR

So in general I think the first option is better in the long run. I can do it but I don't want to clutter this PR even more, that's why I picked the second option for now

Reply:
I totally understand it. We both agree that it needs to have concrete types but that would be a lot of work..

I think that it might be safe as we have a lot of tests and they were already passing so it should be mostly correct I suppose. Could you create a sub-ticket linked to the task to tackle this one separately after everything is done? No need for you to do that, anyone could take on.

Reply:
Sure thing :+1:

Reply:
Done